include/picongpu/plugins/hdf5/HDF5Writer.hpp

```diff
@@ -88,60 +94,284 @@ using namespace splash;
  *
  * Implements the IIOBackend interface.
  */
-class HDF5Writer : public IIOBackend
+class HDF5Writer :
+    public IIOBackend
 {
 public:
+    static std::vector< std::string > splitString(const std::string& input, const std::string& regex) {
+        // passing -1 as the submatch index parameter performs splitting
+        std::regex re(regex);
+        std::sregex_token_iterator
+            first{input.begin(), input.end(), re, -1},
+            last;
+        return {first, last};
+    }
+
+
+    struct Help : public plugins::multi::IHelp
+    {
+        /** creates a instance of ISlave
+         *
+         * @tparam T_Slave type of the interface implementation (must inherit from ISlave)
+         * @param help plugin defined help
+         * @param id index of the plugin, range: [0;help->getNumPlugins())
+         */
+        std::shared_ptr< ISlave > create(
+            std::shared_ptr< IHelp > & help,
+            size_t const id,
+            MappingDesc* cellDescription
+        )
+        {
+            return std::shared_ptr< ISlave >(
+                new HDF5Writer(
+                    help,
+                    id,
+                    cellDescription
+                )
+            );
+        }
+
+        //! periodicity of computing the particle energy
+        plugins::multi::HelpValue< uint32_t > notifyPeriod = {
+            "period",
+            "enable HDF5 IO [for each n-th step]"
+        };
+        plugins::multi::HelpValue< std::string > filter = {
+            "filter",
+            "particle filter: ",
+            "all"
+        };
+
+        plugins::multi::HelpValue< std::string > fileName = {
+            "file",
+            "HDF5 output filename (prefix)"
+        };
+
+        bool isIndependent = false;
+
+        //! string list with all possible particle filters
+        std::string concatenatedFilterNames;
+        std::vector< std::string > allowedFilters;
+
+        template<typename T_TupleVector>
+        struct CreateSpeciesFilter
+        {
+            using type = plugins::misc::SpeciesFilter<
+                typename pmacc::math::CT::At<
+                    T_TupleVector,
+                    bmpl::int_<0>
```

Comment:
we could re-use / refactor the enums and classes from #2368 here if we move them to a general PIConGPU place

Reply:
we can address this later (future PR). all in all it is something like an apply maybe we can write it more boost standard conform.